src/r3d.imageio/r3dinput.cpp

```diff
@@ -0,0 +1,441 @@
+
+    // calculate how many bytes we need
+    sizeNeeded = 16U - (ptr % 16U);
+
+    return buffer + sizeNeeded;
+}
+
+OIIO_PLUGIN_NAMESPACE_BEGIN
+
+#define DBG if (1)
+
+class R3dInput final : public ImageInput {
+public:
+    R3dInput()
+    {
+        initialize();
+        reset();
+    }
+    ~R3dInput() override
+    {
+        close();
+        terminate();
+    }
+    const char* format_name(void) const override { return "r3d"; }
+    int supports(string_view feature) const override
+    {
+        return (feature == "ioproxy");
+    }
+    bool open(const std::string& name, ImageSpec& spec) override;
+    bool open(const std::string& name, ImageSpec& spec,
+              const ImageSpec& config) override;
+    bool seek_subimage(int subimage, int miplevel) override;
+    bool read_native_scanline(int subimage, int miplevel, int y, int z,
+                              void* data) override;
+    void read_frame(int pos);
+    int current_subimage(void) const override
+    {
+        lock_guard lock(*this);
+        return m_subimage;
+    }
+    bool close() override;
+
+    const std::string& filename() const { return m_filename; }
+
+    bool seek(int pos);
+    double fps() const;
+    int64_t time_stamp(int pos) const;
+
+private:
+    std::string m_filename;
+    std::unique_ptr<ImageSpec> m_config;  // Saved copy of configuration spec
+    R3DSDK::Clip* m_clip;
+    R3DSDK::VideoDecodeJob m_job;
+    unsigned char* m_image_buffer;
+    size_t m_adjusted;
+    int m_frames;
+    int m_channels;
+    float m_fps;
+    int m_subimage;
+    int64_t m_nsubimages;
+    int m_last_search_pos;
+    int m_last_decoded_pos;
+    bool m_read_frame;
+    int m_next_scanline;
+    // std::vector<uint16_t> m_pixels;
+
+    void initialize();
+    void reset()
+    {
+        DBG std::cout << "R3dInput::reset()\n";
+
+        ioproxy_clear();
+        m_config.reset();
+        m_clip             = nullptr;
+        m_next_scanline    = 0;
+        m_read_frame       = false;
+        m_subimage         = 0;
+        m_last_search_pos  = 0;
+        m_last_decoded_pos = 0;
+        m_image_buffer     = nullptr;
+        m_adjusted         = 0;
+    }
+
+    void close_file() { reset(); }
+    void terminate();
+};
+
+// Export version number and create function symbols
+OIIO_PLUGIN_EXPORTS_BEGIN
+
+OIIO_EXPORT int r3d_imageio_version = OIIO_PLUGIN_VERSION;
+
+OIIO_EXPORT const char*
+r3d_imageio_library_version()
+{
+    return "R3D";
+}
+
+OIIO_EXPORT ImageInput*
+r3d_input_imageio_create()
+{
+    return new R3dInput;
+}
+
+OIIO_EXPORT const char* r3d_input_extensions[] = { "r3d", nullptr };
+
+OIIO_PLUGIN_EXPORTS_END
+
+
+
+void
+R3dInput::initialize()
+{
+    DBG std::cout << "R3dInput::initialize()\n";
+
+    // initialize SDK
+    // R3DSDK::InitializeStatus init_status = R3DSDK::InitializeSdk(".", OPTION_RED_CUDA);
+    R3DSDK::InitializeStatus init_status
+        = R3DSDK::InitializeSdk("/opt/R3DSDKv8_5_1/Redistributable/linux",
```

Comment:
Should we be worried that this is an absolute path? It is really true that this is the only place, on any system, where it could be installed? What about when not on Linux? Does the SDK exist for other platforms?

Reply:
peter.kovar@Pascal /opt/R3DSDKv8_5_1/Redistributable
€ LANG=en_US ls -lR
.:
total 4
drwxrwxr-x 2 root root  119 Mar  8 15:15 linux
drwxrwxr-x 2 root root   95 Mar  8 15:15 mac
drwxrwxr-x 2 root root 4096 Mar  8 15:15 win

./linux:
total 43300
-rwxr-xr-x 1 root root 11970637 Mar  8 15:01 REDCuda-x64.so
-rwxr-xr-x 1 root root 16306349 Mar  8 15:01 REDDecoder-x64.so
-rwxr-xr-x 1 root root  6068308 Mar  8 14:53 REDOpenCL-x64.so
-rwxrwxr-x 1 root root  4963603 Mar  8 14:01 REDR3D-x64.so
-rwxrwxr-x 1 root root  5008183 Mar  8 14:01 REDR3D-x86.so

./mac:
total 32552
-rwxr-xr-x 1 root root 11544448 Mar  8 14:13 REDDecoder.dylib
-rwxr-xr-x 1 root root  3459216 Mar  8 14:13 REDMetal.dylib
-rwxr-xr-x 1 root root 11112384 Mar  8 14:13 REDOpenCL.dylib
-rwxr-xr-x 1 root root  7198960 Mar  8 14:01 REDR3D.dylib

./win:
total 55052
-rw-r--r-- 1 root root 12414976 Mar  8 15:14 REDCuda-x64.dll
-rw-r--r-- 1 root root  6083584 Mar  8 14:27 REDCuda-x86.dll
-rw-r--r-- 1 root root 15573504 Mar  8 15:14 REDDecoder-x64.dll
-rw-r--r-- 1 root root  4557312 Mar  8 14:30 REDDecoder-x86.dll
-rw-r--r-- 1 root root  5178880 Mar  8 14:35 REDOpenCL-x64.dll
-rw-r--r-- 1 root root  4408320 Mar  8 14:30 REDOpenCL-x86.dll
-rw-r--r-- 1 root root  4258816 Mar  8 14:04 REDR3D-x64.dll
-rw-r--r-- 1 root root  3871744 Mar  8 14:03 REDR3D-x86.dll

Reply:
So in total 5 variants, but maybe they will add macOS AArch64 variant too in not so distant future. I wanted to have something working on Linux x86-64 at first.

Reply:
I think we can certainly figure out programamtically whether to end with linux, mac, win. I'm more concerned about the `/opt/R3DSDKv8_5_1/Redistributable/` part -- not only is the version baked in, but /opt, really, is that the one and only place it could ever be?

This may be ok as a first stab at it, but I wonder if maybe what we eventually want is some kind of search procedure, like we have for finding fonts in imagebufalgo_draw.cpp, where the location could be different for different systems, so we check several likely places, and also have an environment variable that can be used to explicitly give a nonstandard location?

Reply:
What is the best practice in this domain? Sync with R3D so we will not reinvent the wheel.

I would propose following search order:

1. Check of the environment variable and insert its value in the search path.
2. Check and read the runtime configuration file in user home directory.
3. Check the system wide configuration file common to all users.
4. At last try the value obtained at build time.